src/Database/Esqueleto/Internal/Internal.hs

```diff
@@ -1405,26 +1426,105 @@ data Insertion a
 --
 -- @since 2.2.7
-data LockingKind
-    = ForUpdate
+data LockingKind where
+    ForUpdate :: LockingKind
       -- ^ @FOR UPDATE@ syntax.  Supported by MySQL, Oracle and
       -- PostgreSQL.
       --
       -- @since 2.2.7
-    | ForUpdateSkipLocked
+    ForUpdateSkipLocked :: LockingKind
       -- ^ @FOR UPDATE SKIP LOCKED@ syntax.  Supported by MySQL, Oracle and
       -- PostgreSQL.
       --
       -- @since 2.2.7
-    | ForShare
+    ForShare :: LockingKind
       -- ^ @FOR SHARE@ syntax.  Supported by PostgreSQL.
       --
       -- @since 2.2.7
-    | LockInShareMode
+    LockInShareMode :: LockingKind
       -- ^ @LOCK IN SHARE MODE@ syntax.  Supported by MySQL.
       --
       -- @since 2.2.7
 
+-- | Postgres specific locking, used only internally
+--
+-- @since 3.5.8.0
+data PostgresLockingKind where
+  PostgresLockingKind :: PostgresRowLevelLockStrength -> OnLockedBehavior -> PostgresLockingKind
+
+instance Eq PostgresLockingKind where
+  (==) a b = compare a b == EQ
+
+instance Ord PostgresLockingKind where
+  compare (PostgresLockingKind rowstr1 onLockedBehavior1) (PostgresLockingKind rowstr2 onLockedBehavior2) =
+    case compare rowstr1 rowstr2 of
+      EQ -> compare onLockedBehavior1 onLockedBehavior2
+      rowStrCmp -> rowStrCmp
+
+data PostgresRowLevelLockStrength =
+    PostgresForUpdate (Maybe LockingOfClause)
+    | PostgresForShare (Maybe LockingOfClause)
+
+-- | Eq instance of RowLevelLockStrength refers to the equality of the 
+-- strength of lock specifically
+instance Eq PostgresRowLevelLockStrength where
+  (==) a b = compare a b == EQ
+
+-- | Ord instance of RowLevelLockStrength refers to the order of the strength
+-- of lock only. 
+instance Ord PostgresRowLevelLockStrength where
+  compare (PostgresForUpdate _) (PostgresForUpdate _) = EQ
+  compare (PostgresForUpdate _) _  = GT
+  compare _ (PostgresForUpdate _)  = LT
+  compare (PostgresForShare _) (PostgresForShare _)  = EQ
+
+data LockingOfClause where
+  Of :: LockableEntity a => a -> LockingOfClause
+
+data OnLockedBehavior =
+  -- ^ @NOWAIT@ syntax locking behaviour.
+  --  query excutes immediately failing 
+  -- 
+  -- @since 3.5.8.0
+  NoWait
+  -- ^ @SKIP LOCKED@ syntax locking behaviour.
+  --  query skips locked rows
+  -- 
+  -- @since 3.5.8.0
+    | SkipLocked
+    | Wait
+    deriving (Ord, Eq, Show)
+
+
+-- | Lockable entity
+--
+-- Example use:
+--
+-- @
+-- select $ do
+--     (p :& bp) <- from $ 
+--         table @Person
+--         `innerJoin` table @BlogPost
+--             `on` do
+--                 \(p :& bp) -> p ^. PersonId ==. b ^. BlogPostAuthorId
+--     forUpdateOf (p :& b) skipLocked
+--     return p
+-- @
+class LockableEntity a where
+  flattenLockableEntity :: a -> [LockableSqlExpr]
+  makeLockableEntity :: LockableEntity a => IdentInfo -> a -> (TLB.Builder, [PersistValue])
+  makeLockableEntity info lockableEntity =
+    uncommas' $ (\(LockableSqlExpr (ERaw _ f)) -> f Never info) <$> flattenLockableEntity lockableEntity
```

Comment:
If we're not overriding this in any instances, then it can be supplied as a stand-alone function instead of a class member instance. Haven't seen all instances yet here so this may be a redundant comment